src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (expected_ty, &ty::TyRef(_, checked_ty)) => {
+                if expected_ty == &checked_ty.ty.sty {
```

Comment:
using `==` is a bit strict. We might want to use instead `infcx.can_sub(check_ty, expected_ty)`.